opm/autodiff/MSWellHelpers.hpp

```diff
@@ -0,0 +1,195 @@
+/*
+  Copyright 2017 SINTEF Digital, Mathematics and Cybernetics.
+  Copyright 2017 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef OPM_MSWELLHELPERS_HEADER_INCLUDED
+#define OPM_MSWELLHELPERS_HEADER_INCLUDED
+
+#include <opm/common/ErrorMacros.hpp>
+#include <dune/istl/solvers.hh>
+#include <dune/istl/umfpack.hh>
+#include <cmath>
+
+namespace Opm {
+
+namespace mswellhelpers
+{
+
+    // obtain y = D^-1 * x with a direct solver
+    template <typename MatrixType, typename VectorType>
+    VectorType
+    invDXDirect(const MatrixType& D, VectorType x)
```

Comment:
Could `x` be a const reference?

Reply:
Testing result shows no. x will be changed during the solution procedure.